modules/substrate/src/lib.rs

```diff
@@ -311,18 +343,42 @@ impl<T: Trait> BridgeStorage for PalletStorage<T> {
 	type Header = BridgedHeader<T>;
 
 	fn write_header(&mut self, header: &ImportedHeader<BridgedHeader<T>>) {
-		let hash = header.header.hash();
-		<ImportedHeaders<T>>::insert(hash, header);
-	}
+		use core::cmp::Ordering;
+
+		let hash = header.hash();
+		let current_height = header.number();
+		let best_height = <BestHeight<T>>::get();
+
+		match current_height.cmp(&best_height) {
+			Ordering::Equal => {
+				<BestHeaders<T>>::append(hash);
+			}
+			Ordering::Greater => {
+				<BestHeaders<T>>::kill();
+				<BestHeaders<T>>::append(hash);
+				<BestHeight<T>>::put(current_height);
+			}
+			Ordering::Less => {
+				// This is fine. We can still have a valid header, but it might just be on a
+				// different fork and at a lower height than the "best" overall header.
+			}
+		}
 
-	fn best_header(&self) -> ImportedHeader<BridgedHeader<T>> {
-		let hash = <BestHeader<T>>::get();
-		self.header_by_hash(hash)
-			.expect("A header must have been written at genesis, therefore this must always exist")
+		if header.requires_justification {
+			<RequiresJustification<T>>::insert(hash, current_height);
+		} else if <RequiresJustification<T>>::contains_key(hash) {
+			<RequiresJustification<T>>::remove(hash);
+		}
```

Comment:
```suggestion
		} else {
			<RequiresJustification<T>>::remove(hash);
		}
```
To minimize storage access - if the key is not there it's a no-op anyway and at least we don't call into the host twice.